File: iOS/ssabree/ssabree/Core/Repository/PostRepositoryImpl.swift
import Foundation

final class PostRepositoryImpl: PostRepository {
    private let postService: PostService
    private let authDataStore: AuthDataStore

    init(postService: PostService, authDataStore: AuthDataStore) {
        self.postService = postService
        self.authDataStore = authDataStore
    }

    func getPosts(boardId: Int?, keyword: String?, cursor: String?, limit: Int = 20) async -> Result<PagedPostModel, Error> {
        do {
            let response = try await postService.getPosts(boardId: boardId, keyword: keyword, cursor: cursor, limit: limit)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func getHotPosts(cursor: String?, limit: Int = 20) async -> Result<PagedPostModel, Error> {
        do {
            let response = try await postService.getHotPosts(cursor: cursor, limit: limit)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func getPostDetail(postId: Int) async -> Result<PostDetailModel, Error> {
        do {
            let response = try await postService.getPost(id: postId)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func createPost(post: PostCreateInfo) async -> Result<PostModel, Error> {
        do {
            let request = PostCreateRequest(
                title: post.title,
                content: post.content,
                boardId: post.boardId,
                imageUrls: post.images,
                poll: post.poll?.toRequest()
            )
            let response = try await postService.createPost(request)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func updatePost(postId: Int, post: PostUpdateInfo) async -> Result<PostModel, Error> {
        do {
            let request = PostUpdateRequest(title: post.title, content: post.content)
            let response = try await postService.updatePost(id: postId, request)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func deletePost(postId: Int) async -> Result<Void, Error> {
        do {
            try await postService.deletePost(id: postId)
            return .success(())
        } catch {
            return .failure(error)
        }
    }

    func like(postId: Int) async -> Result<PostLikeModel, Error> {
        do {
            let response = try await postService.likePost(postId: postId)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func unlike(postId: Int) async -> Result<PostLikeModel, Error> {
        do {
            let response = try await postService.unlikePost(postId: postId)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func scrap(postId: Int) async -> Result<ScrapModel, Error> {
        do {
            let response = try await postService.scrapPost(postId: postId)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func unscrap(postId: Int) async -> Result<ScrapModel, Error> {
        do {
            let response = try await postService.unscrapPost(postId: postId)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func createComment(postId: Int, comment: CommentCreateInfo) async -> Result<CommentModel, Error> {
        do {
            let request = CommentCreateRequest(content: comment.content)
            let response = try await postService.createComment(postId: postId, request)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func createReply(postId: Int, commentId: Int, reply: ReplyCreateInfo) async -> Result<ReplyModel, Error> {
        do {
            let request = ReplyCreateRequest(content: reply.content)
            let response = try await postService.createReply(postId: postId, commentId: commentId, request)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }

    func vote(postId: Int, vote: VoteInfo) async -> Result<PollModel, Error> {
        do {
            let request = VoteRequest(optionId: vote.optionId)
            let response = try await postService.vote(postId: postId, request)
            return .success(response.toModel())
        } catch {
            return .failure(error)
        }
    }
}

// MARK: - Response to Model Mappers

extension PagedPostResponse {
    func toModel() -> PagedPostModel {
        PagedPostModel(
            data: posts.map { $0.toModel() },
            hasNext: hasNext,
            nextCursor: nextCursor
        )
    }
}

extension PostResponse {
    func toModel() -> PostModel {
        PostModel(
            id: id,
            boardId: boardId,
            boardName: boardName,
            isMine: isMine,
            title: title,
            content: content,
            viewCount: viewCount,
            likeCount: likeCount,
            commentCount: commentCount,
            createdAt: createdAt ?? "",
            imageUrls: imageUrls ?? [],
            isBlinded: isBlinded
        )
    }
}

extension PostDetailResponse {
    func toModel() -> PostDetailModel {
        PostDetailModel(
            createdAt: createdAt ?? "",
            updatedAt: updatedAt ?? "",
            id: id,
            boardId: boardId,
            isMine: isMine,
            title: title,
            content: content,
            isBlinded: isBlinded,
            imageUrls: imageUrls ?? [],
            poll: poll?.toModel(),
            likeCount: likeCount,
            isLiked: isLiked,
            commentCount: commentCount,
            scrapCount: scrapCount,
            isScraped: isScraped,
            comments: comments.map { $0.toModel() }
        )
    }
}

extension CommentResponse {
    func toModel() -> CommentModel {
        CommentModel(
            id: id,
            createdAt: createdAt ?? "",
            content: content,
            likeCount: likeCount,
            isLiked: isLiked,
            isBlinded: isBlinded,
            isDeleted: isDeleted,
            anon: anon?.toModel(),
            replies: replies.map { $0.toModel() }
        )
    }
}

extension ReplyResponse {
    func toModel() -> ReplyModel {
        ReplyModel(
            id: id,
            createdAt: createdAt ?? "",
            content: content,
            likeCount: likeCount,
            isLiked: isLiked,
            isBlinded: isBlinded,
            isDeleted: isDeleted,
            anon: anon?.toModel()
        )
    }
}

extension PollResponse {
    func toModel() -> PollModel {
        PollModel(
            pollId: pollId,
            totalVotes: totalVotes,
            myVotedOptionId: myVotedOptionId,
            options: options.map { $0.toModel() }
        )
    }
}

extension PollOptionResponse {
    func toModel() -> PollOptionModel {
        PollOptionModel(
            optionId: optionId,
            text: text,
            voteCount: voteCount
        )
    }
}

extension PostLikeResponse {
    func toModel() -> PostLikeModel {
        PostLikeModel(
            liked: liked,
            likeCount: likeCount
        )
    }
}

extension ScrapResponse {
    func toModel() -> ScrapModel {
        ScrapModel(
            success: success
        )
    }
}

// MARK: - Info to Request Mappers

extension PollCreateInfo {
    func toRequest() -> PollCreateRequest {
        PollCreateRequest(
            title: title,
            options: options
        )
    }
}

// MARK: - Fake Repository

final class FakePostRepository: PostRepository {
    func getPosts(boardId: Int?, keyword: String?, cursor: String?, limit: Int) async -> Result<PagedPostModel, Error> {
        .success(PagedPostModel(data: [], hasNext: false, nextCursor: nil))
    }

    func getHotPosts(cursor: String?, limit: Int) async -> Result<PagedPostModel, Error> {
        .success(PagedPostModel(data: [], hasNext: false, nextCursor: nil))
    }

    func getPostDetail(postId: Int) async -> Result<PostDetailModel, Error> {
        .success(PostDetailModel(
            createdAt: "",
            updatedAt: "",
            id: 0,
            boardId: 0,
            isMine: false,
            title: "",
            content: "",
            isBlinded: false,
            imageUrls: [],
            poll: nil,
            likeCount: 0,
            isLiked: false,
            commentCount: 0,
            scrapCount: 0,
            isScraped: false,
            comments: []
        ))
    }

    func createPost(post: PostCreateInfo) async -> Result<PostModel, Error> {
        .success(PostModel(id: 0, boardId: 0, boardName: "", isMine: true, title: "", content: "", viewCount: 0, likeCount: 0, commentCount: 0, createdAt: "", imageUrls: [], isBlinded: false))
    }

    func updatePost(postId: Int, post: PostUpdateInfo) async -> Result<PostModel, Error> {
        .success(PostModel(id: 0, boardId: 0, boardName: "", isMine: true, title: "", content: "", viewCount: 0, likeCount: 0, commentCount: 0, createdAt: "", imageUrls: [], isBlinded: false))
    }

    func deletePost(postId: Int) async -> Result<Void, Error> {
        .success(())
    }

    func like(postId: Int) async -> Result<PostLikeModel, Error> {
        .success(PostLikeModel(liked: true, likeCount: 0))
    }

    func unlike(postId: Int) async -> Result<PostLikeModel, Error> {
        .success(PostLikeModel(liked: false, likeCount: 0))
    }

    func scrap(postId: Int) async -> Result<ScrapModel, Error> {
        .success(ScrapModel(success: true))
    }

    func unscrap(postId: Int) async -> Result<ScrapModel, Error> {
        .success(ScrapModel(success: true))
    }

    func createComment(postId: Int, comment: CommentCreateInfo) async -> Result<CommentModel, Error> {
        .success(CommentModel(id: 0, createdAt: "", content: "", likeCount: 0, isLiked: false, isBlinded: false, isDeleted: false, anon: AnonModel(name: "익명", isAuthor: false, isMine: true), replies: []))
    }

    func createReply(postId: Int, commentId: Int, reply: ReplyCreateInfo) async -> Result<ReplyModel, Error> {
        .success(ReplyModel(id: 0, createdAt: "", content: "", likeCount: 0, isLiked: false, isBlinded: false, isDeleted: false, anon: AnonModel(name: "익명", isAuthor: false, isMine: true)))
    }

    func vote(postId: Int, vote: VoteInfo) async -> Result<PollModel, Error> {
        .success(PollModel(pollId: 0, totalVotes: 0, myVotedOptionId: nil, options: []))
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/core/repository/PostRepositoryImpl.kt
package com.ssafy.ssabree.core.repository

import com.ssafy.ssabree.core.datasource.remote.PostService
import com.ssafy.ssabree.core.datasource.remote.model.toCommentCreateRequest
import com.ssafy.ssabree.core.datasource.remote.model.toPostCreateRequest
import com.ssafy.ssabree.core.datasource.remote.model.toPostUpdateRequest
import com.ssafy.ssabree.core.datasource.remote.model.toReplyCreateRequest
import com.ssafy.ssabree.core.datasource.remote.model.toVoteRequest
import com.ssafy.ssabree.core.repository.model.CommentCreateInfo
import com.ssafy.ssabree.core.repository.model.CommentModel
import com.ssafy.ssabree.core.repository.model.PostCreateInfo
import com.ssafy.ssabree.core.repository.model.PostDetailModel
import com.ssafy.ssabree.core.repository.model.PostLikeModel
import com.ssafy.ssabree.core.repository.model.PostModel
import com.ssafy.ssabree.core.repository.model.PagedPostModel
import com.ssafy.ssabree.core.repository.model.PostUpdateInfo
import com.ssafy.ssabree.core.repository.model.PollModel
import com.ssafy.ssabree.core.repository.model.ReplyCreateInfo
import com.ssafy.ssabree.core.repository.model.ReplyModel
import com.ssafy.ssabree.core.repository.model.ScrapModel
import com.ssafy.ssabree.core.repository.model.VoteInfo
import com.ssafy.ssabree.core.repository.model.toModel
import com.ssafy.ssabree.core.utils.RetrofitClient

class PostRepositoryImpl : PostRepository {

    private val postService = RetrofitClient.instance.create(PostService::class.java)

    override suspend fun getPosts(
        boardId: Long?,
        keyword: String?,
        cursor: String?,
        limit: Int
    ): Result<PagedPostModel> {
        return runCatching {
            postService.getPosts(
                boardId = boardId,
                keyword = keyword,
                cursor = cursor,
                limit = limit
            ).toModel()
        }
    }

    override suspend fun getHotPosts(
        cursor: String?,
        limit: Int
    ): Result<PagedPostModel> {
        return runCatching {
            postService.getHotPosts(
                cursor = cursor,
                limit = limit
            ).toModel()
        }
    }

    override suspend fun getPostDetail(id: Long): Result<PostDetailModel> {
        return runCatching {
            postService.getPost(id).toModel()
        }
    }

    override suspend fun createPost(info: PostCreateInfo): Result<PostModel> {
        return runCatching {
            postService.createPost(info.toPostCreateRequest()).toModel()
        }
    }

    override suspend fun updatePost(id: Long, info: PostUpdateInfo): Result<PostModel> {
        return runCatching {
            postService.updatePost(id, info.toPostUpdateRequest()).toModel()
        }
    }

    override suspend fun deletePost(id: Long): Result<Unit> {
        return runCatching {
            postService.deletePost(id)
        }
    }

    override suspend fun createComment(
        postId: Long,
        info: CommentCreateInfo
    ): Result<CommentModel> {
        return runCatching {
            postService.createComment(postId, info.toCommentCreateRequest()).toModel()
        }
    }

    override suspend fun createReply(
        postId: Long,
        commentId: Long,
        info: ReplyCreateInfo
    ): Result<ReplyModel> {
        return runCatching {
            postService.createReply(postId, commentId, info.toReplyCreateRequest()).toModel()
        }
    }

    override suspend fun vote(postId: Long, info: VoteInfo): Result<PollModel> {
        return runCatching {
            postService.vote(postId, info.toVoteRequest()).toModel()
        }
    }

    override suspend fun like(postId: Long): Result<PostLikeModel> {
        return runCatching {
            postService.like(postId).toModel()
        }
    }

    override suspend fun unlike(postId: Long): Result<PostLikeModel> {
        return runCatching {
            postService.unlike(postId).toModel()
        }
    }

    override suspend fun scrap(postId: Long): Result<ScrapModel> {
        return runCatching {
            postService.scrap(postId).toModel()
        }
    }

    override suspend fun unscrap(postId: Long): Result<ScrapModel> {
        return runCatching {
            postService.unscrap(postId).toModel()
        }
    }
}
